### ISSUE TITLE
An exception that is thrown logs an object that has been deleted

### ISSUE THREAD

**Issue by xeroc** [bitshares/bitshares-core#383]
Credits to @dnotestein and team.

In [this line](https://github.com/bitshares/bitshares-core/blob/master/libraries/chain/db_block.cpp#L295), we throw an exception with `(proposal)` but we might have deleted `proposal` in [a previous line](https://github.com/bitshares/bitshares-core/blob/master/libraries/chain/db_block.cpp#L274).

**Comment by pmconrad** [comment]
Nice catch, however...

I think in practice this shouldn't be a problem, because the exception would have to be thrown *after* the remove(proposal) call, which leaves only the session.merge() and the move assignment as possible causes, both of which are unlikely to throw.

OTOH the typical case is that one of the apply_operation calls before the remove throws, and in that case the current proposal is valuable information and should be logged.

**Comment by abitmore** [comment]
To be safe we can change the code to something like this:
```
try{
   the_code_before_remove();
} FC_CAPTURE_AND_RETHROW( (proposal) )

try{
   remove(*proposal);
   merge();
}FC_CAPTURE_AND_RETHROW()
```

**Comment by xeroc** [comment]
Unfortunately, I have no clue how to properly deal with `session` which is defined in the `try` section.
Can that definition be moved outside the inner `try` block so that it is available in a second `try` block?
Maybe someone can show me how to do it, I'll glady modify the PR.

**Comment by abitmore** [comment]
Won't fix as discussed in #739.
